File: src/main/java/me/Geek/Libs/DataBase/mysql.java
package me.Geek.Libs.DataBase;

import com.zaxxer.hikari.HikariDataSource;
import me.Geek.Configuration.ConfigManager;

import java.sql.Connection;
import java.sql.SQLException;
import java.sql.Statement;

/**
 * 作者: 老廖
 * 时间: 2022/7/23
 **/
public class mysql extends DataSub{
    private HikariDataSource MYSQL;

    public mysql() {
        super();
    }
    @Override
    public Connection getConnection() throws SQLException {
        return MYSQL.getConnection();
    }

    @Override
    public void onLoad() {
        final String MysqlUrl = "jdbc:mysql://" + ConfigManager.MYSQL_HOST + ":" + ConfigManager.MYSQL_PORT + "/" + ConfigManager.MYSQL_DATABASE + ConfigManager.MYSQL_PARAMS;
        MYSQL = new HikariDataSource();
        MYSQL.setJdbcUrl(MysqlUrl);
        MYSQL.setUsername(ConfigManager.MYSQL_USERNAME);
        MYSQL.setPassword(ConfigManager.MYSQL_PASSWORD);
        // 设置驱动
        try {
            MYSQL.setDriverClassName("com.mysql.cj.jdbc.Driver");
        } catch (RuntimeException | NoClassDefFoundError e) {
            MYSQL.setDriverClassName("com.mysql.jdbc.Driver");
        }
        MYSQL.setMaximumPoolSize(ConfigManager.MAXIMUM_POOL_SIZE);
        MYSQL.setMinimumIdle(ConfigManager.MINIMUM_IDLE);
        MYSQL.setMaxLifetime(ConfigManager.MAXIMUM_LIFETIME);
        MYSQL.setKeepaliveTime(ConfigManager.KEEPALIVE_TIME);
        MYSQL.setConnectionTimeout(ConfigManager.CONNECTION_TIMEOUT);
        MYSQL.setPoolName("GeekMail-MYSQL");
        createMysqlTables();
    }

    @Override
    public void onStop() {
        if (MYSQL != null) MYSQL.close();
    }

    private void createMysqlTables() {
        try (Connection connection = getConnection()) {
            try (Statement statement = connection.createStatement()){
                statement.execute("CREATE TABLE IF NOT EXISTS `mail_data` (" +
                        " `id` INT(80) NOT NULL AUTO_INCREMENT, " +
                        " `mail_id` CHAR(36) NOT NULL, " +
                        " `state` VARCHAR(256) NOT NULL, " +
                        " `type` text NOT NULL, " +
                        " `sender` CHAR(36) NOT NULL, " +
                        " `target` CHAR(36) NOT NULL, " +
                        " `title` text NOT NULL, " +
                        " `text` longtext NOT NULL, " +
                        " `money` double NOT NULL DEFAULT '0', " +
                        " `points` integer NOT NULL DEFAULT '0', " +
                        " `exp` integer NOT NULL DEFAULT '0', " +
                        " `item` longtext NOT NULL, " +
                        " `commands` longtext NOT NULL, " +
                        " PRIMARY KEY (`id`))ENGINE=InnoDB DEFAULT CHARSET=utf8;");
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }
}

File: src/main/kotlin/me/geek/mail/common/data/Mysql.kt
package me.geek.mail.common.data

import com.zaxxer.hikari.HikariDataSource
import me.geek.mail.common.data.sub.DataSub
import me.geek.mail.common.data.sub.action
import me.geek.mail.common.data.sub.use
import me.geek.mail.modules.settings.SetTings
import java.lang.RuntimeException
import java.sql.Connection
import java.sql.SQLException

/**
 * 作者: 老廖
 * 时间: 2022/7/23
 */
class Mysql : DataSub() {
    private var MYSQL: HikariDataSource = HikariDataSource()
    private val set = SetTings.StorageDate

    override val connection: Connection
        get() = MYSQL.connection

    override fun onLoad() {
        val MysqlUrl =
            "jdbc:mysql://" + set.mysql.host + ":" + set.mysql.port + "/" + set.mysql.database + set.mysql.params
        MYSQL.jdbcUrl = MysqlUrl
        MYSQL.username = set.mysql.username
        MYSQL.password = set.mysql.password
        // 设置驱动
        try {
            MYSQL.driverClassName = "com.mysql.cj.jdbc.Driver"
        } catch (e: RuntimeException) {
            MYSQL.driverClassName = "com.mysql.jdbc.Driver"
        } catch (e: NoClassDefFoundError) {
            MYSQL.driverClassName = "com.mysql.jdbc.Driver"
        }
        MYSQL.maximumPoolSize = set.hikari_settings.maximum_pool_size
        MYSQL.minimumIdle = set.hikari_settings.minimum_idle
        MYSQL.maxLifetime = set.hikari_settings.maximum_lifetime.toLong()
        MYSQL.keepaliveTime = set.hikari_settings.keepalive_time.toLong()
        MYSQL.connectionTimeout = set.hikari_settings.connection_timeout.toLong()
        MYSQL.poolName = "GeekMail-MYSQL"
        createMysqlTables()
    }

    override fun onStop() {
        MYSQL.close()
    }

    private fun createMysqlTables() {
        try {
            connection.use {
                this.createStatement().action { statement ->
                    statement.addBatch(
                        "CREATE TABLE IF NOT EXISTS `mail_player_data` (" +
                                " `uuid` CHAR(36) NOT NULL UNIQUE," +
                                " `name` VARCHAR(16) NOT NULL," +
                                " `mail` CHAR(254) NOT NULL," +
                                " `one_join` CHAR(5) NOT NULL," +
                                "PRIMARY KEY (`uuid`)" +
                                ");"
                    )
                    statement.addBatch(
                        "CREATE TABLE IF NOT EXISTS `maildata` (" +
                                " `id` integer NOT NULL AUTO_INCREMENT, " +
                                " `mail_id` CHAR(36) NOT NULL UNIQUE, " +
                                " `state` VARCHAR(256) NOT NULL, " +
                                " `type` text NOT NULL, " +
                                " `sender` CHAR(36) NOT NULL, " +
                                " `target` CHAR(36) NOT NULL, " +
                                " `title` text NOT NULL, " +
                                " `text` longtext NOT NULL, " +
                                " `additional` text NOT NULL, " +
                                " `item` longtext NOT NULL, " +
                                " `commands` longtext NOT NULL, " +
                                " `sendertime` BIGINT(20) NOT NULL, " +
                                " `gettime` BIGINT(20) NOT NULL, " +
                                " PRIMARY KEY (`id`,`target`)" +
                                ");"
                    )
                    statement.executeBatch()
                }
            }
        } catch (e: SQLException) {
            e.printStackTrace()
        }
    }
}